fix initialization on VF

ajitkhaparde opened the pull request:
1) Remove PF specific code in VF initialization path.
2) Add VF device id for Whitney+ VF support

stephen-hurd commented on drivers/net/bnxt/bnxt_cpr.c:
```diff
@@ -157,7 +157,8 @@ int bnxt_alloc_def_cp_ring(struct bnxt *bp)
 	bp->grp_info[0].cp_fw_ring_id = cp_ring->fw_ring_id;
-	rc = bnxt_hwrm_func_cfg_def_cp(bp);
+	if (!BNXT_VF(bp))
+		rc = bnxt_hwrm_func_cfg_def_cp(bp);
```
A default completion ring isn't just for the PF, the function should be modified to work with the VFs instead.

ajitkhaparde replied:
But is a VF allowed to call with HWRM_FUNC_CFG_INPUT_ENABLES_ASYNC_EVENT_CR ?

stephen-hurd replied:
Async events are used for things like link state changes, so I certainly hope so.

stephen-hurd commented on drivers/net/bnxt/bnxt_ethdev.c:
```diff
@@ -1422,8 +1424,7 @@ bnxt_dev_init(struct rte_eth_dev *eth_dev)
 			}
-		}
-		else {
+		} else if (!BNXT_VF(bp)) {
```
How can you get here when !BNXT_VF(bp) since it's inside a BNXT_PF(bp) block?

ajitkhaparde replied:
This is stale. I think we can remove that since I have added the VF device ID.

ajitkhaparde commented:
I looked at the comment in the HWRM.
Is if we don't send this ioctl for a VF, we will still get it on the
default CP ring - just like old days.
Unless I understood it wrong.

uint16_t async_event_cr;

        /*

         * ID of the target completion ring for receiving asynchronous

         * event completions. If this field is not valid, then the HWRM

         * shall use the default completion ring of the function that is

         * being configured as the target completion ring for providing

         * any asynchronous event completions for that function. If this

         * field is valid, then the HWRM shall use the completion ring

         * identified by this ID as the target completion ring for

         * providing any asynchronous event completions for the function

         * that is being configured.

On Wed, May 24, 2017 at 5:25 PM, Stephen Hurd <notifications@github.com>
wrote:

> *@stephen-hurd* commented on this pull request.
> ------------------------------
>
> In drivers/net/bnxt/bnxt_cpr.c
> <https://github.com/Broadcom/dpdk/pull/58#discussion_r118379919>:
>
> > @@ -157,7 +157,8 @@ int bnxt_alloc_def_cp_ring(struct bnxt *bp)
>  	cpr->cp_doorbell = bp->pdev->mem_resource[2].addr;
>  	B_CP_DB_DISARM(cpr);
>  	bp->grp_info[0].cp_fw_ring_id = cp_ring->fw_ring_id;
> -	rc = bnxt_hwrm_func_cfg_def_cp(bp);
> +	if (!BNXT_VF(bp))
> +		rc = bnxt_hwrm_func_cfg_def_cp(bp);
>
> Async events are used for things like link state changes, so I certainly
> hope so.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Broadcom/dpdk/pull/58#discussion_r118379919>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AXqwrGzVAvhfsHQEa8iAgRLcInn9k5ujks5r9K5CgaJpZM4NlSMY>
> .
>

stephen-hurd commented:
But that doesn't add up to a reason to call it for the PF, but not for the
VF... it sounds more like a reason to delete that function completely.

On Wed, May 24, 2017 at 3:30 PM, Ajit Khaparde <notifications@github.com>
wrote:

> I looked at the comment in the HWRM.
> Is if we don't send this ioctl for a VF, we will still get it on the
> default CP ring - just like old days.
> Unless I understood it wrong.
>
> uint16_t async_event_cr;
>
> /*
>
> * ID of the target completion ring for receiving asynchronous
>
> * event completions. If this field is not valid, then the HWRM
>
> * shall use the default completion ring of the function that is
>
> * being configured as the target completion ring for providing
>
> * any asynchronous event completions for that function. If this
>
> * field is valid, then the HWRM shall use the completion ring
>
> * identified by this ID as the target completion ring for
>
> * providing any asynchronous event completions for the function
>
> * that is being configured.
>
> On Wed, May 24, 2017 at 5:25 PM, Stephen Hurd <notifications@github.com>
> wrote:
>
> > *@stephen-hurd* commented on this pull request.
> > ------------------------------
> >
> > In drivers/net/bnxt/bnxt_cpr.c
> > <https://github.com/Broadcom/dpdk/pull/58#discussion_r118379919>:
> >
> > > @@ -157,7 +157,8 @@ int bnxt_alloc_def_cp_ring(struct bnxt *bp)
> > cpr->cp_doorbell = bp->pdev->mem_resource[2].addr;
> > B_CP_DB_DISARM(cpr);
> > bp->grp_info[0].cp_fw_ring_id = cp_ring->fw_ring_id;
> > - rc = bnxt_hwrm_func_cfg_def_cp(bp);
> > + if (!BNXT_VF(bp))
> > + rc = bnxt_hwrm_func_cfg_def_cp(bp);
> >
> > Async events are used for things like link state changes, so I certainly
> > hope so.
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/Broadcom/dpdk/pull/58#discussion_r118379919>, or
> mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AXqwrGzVAvhfsHQEa8iAgRLcInn9k5ujks5r9K5CgaJpZM4NlSMY>
> > .
> >
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/Broadcom/dpdk/pull/58#issuecomment-303870220>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ASsTfaISNG8PcfM-FMIFKOy36cxeTos_ks5r9K9sgaJpZM4NlSMY>
> .
>



-- 
Stephen Hurd
949-926-8039
stephen.hurd@broadcom.com

ajitkhaparde commented:
Which means the changes that you pushed for "Cp ring update" push request
#34
should be abandoned? Or is it required for VF forwarding? Because I think
that was the motivation at that time?

On Wed, May 24, 2017 at 5:36 PM, Stephen Hurd <notifications@github.com>
wrote:

> But that doesn't add up to a reason to call it for the PF, but not for the
> VF... it sounds more like a reason to delete that function completely.
>
> On Wed, May 24, 2017 at 3:30 PM, Ajit Khaparde <notifications@github.com>
> wrote:
>
> > I looked at the comment in the HWRM.
> > Is if we don't send this ioctl for a VF, we will still get it on the
> > default CP ring - just like old days.
> > Unless I understood it wrong.
> >
> > uint16_t async_event_cr;
> >
> > /*
> >
> > * ID of the target completion ring for receiving asynchronous
> >
> > * event completions. If this field is not valid, then the HWRM
> >
> > * shall use the default completion ring of the function that is
> >
> > * being configured as the target completion ring for providing
> >
> > * any asynchronous event completions for that function. If this
> >
> > * field is valid, then the HWRM shall use the completion ring
> >
> > * identified by this ID as the target completion ring for
> >
> > * providing any asynchronous event completions for the function
> >
> > * that is being configured.
> >
> > On Wed, May 24, 2017 at 5:25 PM, Stephen Hurd <notifications@github.com>
> > wrote:
> >
> > > *@stephen-hurd* commented on this pull request.
> > > ------------------------------
> > >
> > > In drivers/net/bnxt/bnxt_cpr.c
> > > <https://github.com/Broadcom/dpdk/pull/58#discussion_r118379919>:
> > >
> > > > @@ -157,7 +157,8 @@ int bnxt_alloc_def_cp_ring(struct bnxt *bp)
> > > cpr->cp_doorbell = bp->pdev->mem_resource[2].addr;
> > > B_CP_DB_DISARM(cpr);
> > > bp->grp_info[0].cp_fw_ring_id = cp_ring->fw_ring_id;
> > > - rc = bnxt_hwrm_func_cfg_def_cp(bp);
> > > + if (!BNXT_VF(bp))
> > > + rc = bnxt_hwrm_func_cfg_def_cp(bp);
> > >
> > > Async events are used for things like link state changes, so I
> certainly
> > > hope so.
> > >
> > > —
> > > You are receiving this because you authored the thread.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/Broadcom/dpdk/pull/58#discussion_r118379919>, or
> > mute
> > > the thread
> > > <https://github.com/notifications/unsubscribe-auth/
> > AXqwrGzVAvhfsHQEa8iAgRLcInn9k5ujks5r9K5CgaJpZM4NlSMY>
> > > .
> > >
> >
> > —
> > You are receiving this because your review was requested.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/Broadcom/dpdk/pull/58#issuecomment-303870220>, or
> mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/ASsTfaISNG8PcfM-
> FMIFKOy36cxeTos_ks5r9K9sgaJpZM4NlSMY>
> > .
> >
>
>
>
> --
> Stephen Hurd
> 949-926-8039 <(949)%20926-8039>
> stephen.hurd@broadcom.com
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Broadcom/dpdk/pull/58#issuecomment-303871437>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AXqwrPVw5lMbklG7Cdl2NZ3AcB-HHIKNks5r9LD4gaJpZM4NlSMY>
> .
>

stephen-hurd commented:
Before that commit, the default completion ring was completely broken, so no, please don't abandon the whole change.  :-)

Looks like at the time I felt it was useful to explicitly set the async completion ring.  If the driver still works in a VF without that call, it seems I was mistaken.

ajitkhaparde commented:
Yes. That's how I remembered the change. It's just for this conversation, I
dug the push id. :-)

I have done limited testing on VF, BTW.
Without that check in that function, the initialization will fail for a VF.
After that change, I can do I/O.

For rest of the testing, I want the QA to give a GO no-GO.
And then we can decide on the necessary code change?

On Wed, May 24, 2017 at 6:01 PM, Stephen Hurd <notifications@github.com>
wrote:

> Before that commit, the default completion ring was completely broken, so
> no, please don't abandon the whole change. :-)
>
> Looks like at the time I felt it was useful to explicitly set the async
> completion ring. If the driver still works in a VF without that call, it
> seems I was mistaken.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Broadcom/dpdk/pull/58#issuecomment-303876289>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AXqwrI3o7HlbyxLEHWTr-lxbXusuGu4aks5r9LbUgaJpZM4NlSMY>
> .
>

stephen-hurd commented:
A correct fix would likely be for that function to call hwrm_func_vf_cfg on
VFs... either in that function or as a
separate bnxt_hwrm_func_vf_cfg_def_cp() function 9or remove the function
and the call to it).

Having QA test code we don't plan to ship seems like a waste of what little
of their time we do get.

On Wed, May 24, 2017 at 4:06 PM, Ajit Khaparde <notifications@github.com>
wrote:

> Yes. That's how I remembered the change. It's just for this conversation, I
> dug the push id. :-)
>
> I have done limited testing on VF, BTW.
> Without that check in that function, the initialization will fail for a VF.
> After that change, I can do I/O.
>
> For rest of the testing, I want the QA to give a GO no-GO.
> And then we can decide on the necessary code change?
>
> On Wed, May 24, 2017 at 6:01 PM, Stephen Hurd <notifications@github.com>
> wrote:
>
> > Before that commit, the default completion ring was completely broken, so
> > no, please don't abandon the whole change. :-)
> >
> > Looks like at the time I felt it was useful to explicitly set the async
> > completion ring. If the driver still works in a VF without that call, it
> > seems I was mistaken.
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/Broadcom/dpdk/pull/58#issuecomment-303876289>, or
> mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/AXqwrI3o7HlbyxLEHWTr-
> lxbXusuGu4aks5r9LbUgaJpZM4NlSMY>
> > .
> >
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/Broadcom/dpdk/pull/58#issuecomment-303877253>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ASsTfYoJ7x13rEzw88FErL9Nf_wcKTvNks5r9LfmgaJpZM4NlSMY>
> .
>



-- 
Stephen Hurd
949-926-8039
stephen.hurd@broadcom.com

ajitkhaparde commented:
the firmware is checking for pf_fid == caller_fid.
Will that work for a VF? This ASYNC_EVENT_CR registration seems to have been
written for a PF?


        if (cmd->enables & HWRM_FUNC_CFG_INPUT_ENABLES_ASYNC_EVENT_CR) {
            owner = &CDEV.cmpl_rings[cmd->async_event_cr];
            if ((owner->flags & CFW_HW_RESOURCE_OWNER_FLAGS_ALLOCATED) &&
                owner->pf_fid == caller_fid)
                CDEV.async_cmpl_pf[pf_id] = cmd->async_event_cr;
        }

On Wed, May 24, 2017 at 6:11 PM, Stephen Hurd <notifications@github.com>
wrote:

> A correct fix would likely be for that function to call hwrm_func_vf_cfg on
> VFs... either in that function or as a
> separate bnxt_hwrm_func_vf_cfg_def_cp() function 9or remove the function
> and the call to it).
>
> Having QA test code we don't plan to ship seems like a waste of what little
> of their time we do get.
>
> On Wed, May 24, 2017 at 4:06 PM, Ajit Khaparde <notifications@github.com>
> wrote:
>
> > Yes. That's how I remembered the change. It's just for this
> conversation, I
> > dug the push id. :-)
> >
> > I have done limited testing on VF, BTW.
> > Without that check in that function, the initialization will fail for a
> VF.
> > After that change, I can do I/O.
> >
> > For rest of the testing, I want the QA to give a GO no-GO.
> > And then we can decide on the necessary code change?
> >
> > On Wed, May 24, 2017 at 6:01 PM, Stephen Hurd <notifications@github.com>
> > wrote:
> >
> > > Before that commit, the default completion ring was completely broken,
> so
> > > no, please don't abandon the whole change. :-)
> > >
> > > Looks like at the time I felt it was useful to explicitly set the async
> > > completion ring. If the driver still works in a VF without that call,
> it
> > > seems I was mistaken.
> > >
> > > —
> > > You are receiving this because you authored the thread.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/Broadcom/dpdk/pull/58#issuecomment-303876289>, or
> > mute
> > > the thread
> > > <https://github.com/notifications/unsubscribe-
> auth/AXqwrI3o7HlbyxLEHWTr-
> > lxbXusuGu4aks5r9LbUgaJpZM4NlSMY>
> > > .
> > >
> >
> > —
> > You are receiving this because your review was requested.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/Broadcom/dpdk/pull/58#issuecomment-303877253>, or
> mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/
> ASsTfYoJ7x13rEzw88FErL9Nf_wcKTvNks5r9LfmgaJpZM4NlSMY>
> > .
> >
>
>
>
> --
> Stephen Hurd
> 949-926-8039 <(949)%20926-8039>
> stephen.hurd@broadcom.com
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Broadcom/dpdk/pull/58#issuecomment-303878250>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AXqwrIwFaDsHXm-w1aJeYbRuEc0FS1txks5r9LkhgaJpZM4NlSMY>
> .
>

stephen-hurd commented:
VFs will need to call FUNC_VF_CFG instead of FUNC_CFG.

On Wed, May 24, 2017 at 4:20 PM, Ajit Khaparde <notifications@github.com>
wrote:

> the firmware is checking for pf_fid == caller_fid.
> Will that work for a VF? This ASYNC_EVENT_CR registration seems to have
> been
> written for a PF?
>
>
> if (cmd->enables & HWRM_FUNC_CFG_INPUT_ENABLES_ASYNC_EVENT_CR) {
> owner = &CDEV.cmpl_rings[cmd->async_event_cr];
> if ((owner->flags & CFW_HW_RESOURCE_OWNER_FLAGS_ALLOCATED) &&
> owner->pf_fid == caller_fid)
> CDEV.async_cmpl_pf[pf_id] = cmd->async_event_cr;
> }
>
> On Wed, May 24, 2017 at 6:11 PM, Stephen Hurd <notifications@github.com>
>
> wrote:
>
> > A correct fix would likely be for that function to call hwrm_func_vf_cfg
> on
> > VFs... either in that function or as a
> > separate bnxt_hwrm_func_vf_cfg_def_cp() function 9or remove the function
> > and the call to it).
> >
> > Having QA test code we don't plan to ship seems like a waste of what
> little
> > of their time we do get.
> >
> > On Wed, May 24, 2017 at 4:06 PM, Ajit Khaparde <notifications@github.com
> >
> > wrote:
> >
> > > Yes. That's how I remembered the change. It's just for this
> > conversation, I
> > > dug the push id. :-)
> > >
> > > I have done limited testing on VF, BTW.
> > > Without that check in that function, the initialization will fail for a
> > VF.
> > > After that change, I can do I/O.
> > >
> > > For rest of the testing, I want the QA to give a GO no-GO.
> > > And then we can decide on the necessary code change?
> > >
> > > On Wed, May 24, 2017 at 6:01 PM, Stephen Hurd <
> notifications@github.com>
> > > wrote:
> > >
> > > > Before that commit, the default completion ring was completely
> broken,
> > so
> > > > no, please don't abandon the whole change. :-)
> > > >
> > > > Looks like at the time I felt it was useful to explicitly set the
> async
> > > > completion ring. If the driver still works in a VF without that call,
> > it
> > > > seems I was mistaken.
> > > >
> > > > —
> > > > You are receiving this because you authored the thread.
> > > > Reply to this email directly, view it on GitHub
> > > > <https://github.com/Broadcom/dpdk/pull/58#issuecomment-303876289>,
> or
> > > mute
> > > > the thread
> > > > <https://github.com/notifications/unsubscribe-
> > auth/AXqwrI3o7HlbyxLEHWTr-
> > > lxbXusuGu4aks5r9LbUgaJpZM4NlSMY>
> > > > .
> > > >
> > >
> > > —
> > > You are receiving this because your review was requested.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/Broadcom/dpdk/pull/58#issuecomment-303877253>, or
> > mute
> > > the thread
> > > <https://github.com/notifications/unsubscribe-auth/
> > ASsTfYoJ7x13rEzw88FErL9Nf_wcKTvNks5r9LfmgaJpZM4NlSMY>
> > > .
> > >
> >
> >
> >
> > --
> > Stephen Hurd
> > 949-926-8039 <(949)%20926-8039> <(949)%20926-8039>
> > stephen.hurd@broadcom.com
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/Broadcom/dpdk/pull/58#issuecomment-303878250>, or
> mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/AXqwrIwFaDsHXm-
> w1aJeYbRuEc0FS1txks5r9LkhgaJpZM4NlSMY>
> > .
> >
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/Broadcom/dpdk/pull/58#issuecomment-303879692>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ASsTfRku5fEmsfyLFqk5HE7ICYK6pPPQks5r9LswgaJpZM4NlSMY>
> .
>



-- 
Stephen Hurd
949-926-8039
stephen.hurd@broadcom.com

ajitkhaparde committed:
bnxt: fix initialization issues on VF

We are calling some PF specific code in the VF initialization path.
This patch fixes the issue by calling VF specific HWRM function.

Change-Id: I8c0c48d320a8c64b977a240ab85659b8f721b855
Signed-off-by: Ajit Khaparde <ajit.khaparde@broadcom.com>

ajitkhaparde committed:
bnxt: add PCI device id for Whitney+ VF support

Add PCI device id to support Whitney+ VF.
BROADCOM_DEV_ID_57414_VF 0x16c1

Change-Id: I1550441ba4f8cb9fab57035e6a4c4d3640bc3412
Signed-off-by: Ajit Khaparde <ajit.khaparde@broadcom.com>

stephen-hurd reviewed (approved):
Looks good.
